### ISSUE TITLE
Assertion Error when using VertexAIEmbeddings with faiss vectorstore

### ISSUE THREAD

**Issue by bhanu-pappala** [langchain-ai/langchain#5262]
```
~\Anaconda3\lib\site-packages\langchain\memory\vectorstore.py in save_context(self, inputs, outputs)
     67         """Save context from this conversation to buffer."""
     68         documents = self._form_documents(inputs, outputs)
---> 69         self.retriever.add_documents(documents)
     70 
     71     def clear(self) -> None:

~\Anaconda3\lib\site-packages\langchain\vectorstores\base.py in add_documents(self, documents, **kwargs)
    413     def add_documents(self, documents: List[Document], **kwargs: Any) -> List[str]:
    414         """Add documents to vectorstore."""
--> 415         return self.vectorstore.add_documents(documents, **kwargs)
    416 
    417     async def aadd_documents(

~\Anaconda3\lib\site-packages\langchain\vectorstores\base.py in add_documents(self, documents, **kwargs)
     60         texts = [doc.page_content for doc in documents]
     61         metadatas = [doc.metadata for doc in documents]
---> 62         return self.add_texts(texts, metadatas, **kwargs)
     63 
     64     async def aadd_documents(

~\Anaconda3\lib\site-packages\langchain\vectorstores\faiss.py in add_texts(self, texts, metadatas, ids, **kwargs)
    150         # Embed and create the documents.
    151         embeddings = [self.embedding_function(text) for text in texts]
--> 152         return self.__add(texts, embeddings, metadatas=metadatas, ids=ids, **kwargs)
    153 
    154     def add_embeddings(

~\Anaconda3\lib\site-packages\langchain\vectorstores\faiss.py in __add(self, texts, embeddings, metadatas, ids, **kwargs)
    117         if self._normalize_L2:
    118             faiss.normalize_L2(vector)
--> 119         self.index.add(vector)
    120         # Get list of index, id, and docs.
    121         full_info = [(starting_len + i, ids[i], doc) for i, doc in enumerate(documents)]

~\Anaconda3\lib\site-packages\faiss\class_wrappers.py in replacement_add(self, x)
    226 
    227         n, d = x.shape
--> 228         assert d == self.d
    229         x = np.ascontiguousarray(x, dtype='float32')
    230         self.add_c(n, swig_ptr(x))

AssertionError: 
```

**Comment by dev2049** [comment]
could you provide code snippet to reproduce

**Comment by bhanu-pappala** [comment]
> could you provide code snippet to reproduce

I followed the [documentation](https://python.langchain.com/en/latest/modules/memory/types/vectorstore_retriever_memory.html). In place of openaiembeddings, I specified VertexAIEmbeddings

**Comment by jacquesmats** [comment]
Same error here. What is causing? The empty "AssertionError: " is very disappointing

**Comment by risingsunomi** [comment]
I am also still getting this issue. A snippet (sorry could get the code to display right in this so used quotes)

> self.vectorstore = FAISS(
            VertexAIEmbeddings().embed_query,
            faiss.IndexFlatL2(),
            InMemoryDocstore({}),
            {},
        )
        retriever = self.vectorstore.as_retriever(search_kwargs=dict(k=1))
        self.memory = VectorStoreRetrieverMemory(
            retriever=retriever
        )
        self.conversation = ConversationChain(
            llm=self.llm,
            prompt=self.instruction_prompt_template,
            memory=self.memory,
            verbose=True,
        )

**Comment by risingsunomi** [comment]
Ok I got past the error by setting the faiss.IndexFlatL2 embedding size to 768 but ran into a empty search error with "KeyError 0" which I think is because I didn't put anything in memory. I didn't think that was a requirement from looking at the tutorials.

Apparently you have to do a save_context for this to work with conversationchain. This seems like a bad thing where if you have nothing for save_context then you are out of luck and have to make up something.

Again the solution seems to be

- Set embedding size to 768
- Use save_context right after with self.memory before calling ConversationChain

**Comment by adityak714** [comment]
It was a simpler fix for me than I thought. For context, I was using the Gemini Embedding model. While defining `faiss.IndexFlatL2`, I defined it as `index=faiss.IndexFlatL2(len(embeddings))`([docs](https://python.langchain.com/docs/integrations/vectorstores/faiss/)), but I did not apply it correctly. The demo in the documentation embedded only one string to give an example, hence as there was only 1 chunk present, the length returned would be correct (3,072 for Gemini). But in my use-case, I had multiple split text chunks.

If you have a collection of text chunks that are embedded, and one simply gets the length of the list, it returns the number of text chunks (not what we want). We want the embedding dimension. 

You get that by inspecting ***one*** of the chunks, and getting the length of that (by doing e.g. `len(embeddings[0])`). The embedding dimension you find from that is to be entered in `faiss.IndexFlatL2(...)`. Hope that helps whoever else might be scratching their heads!
